devices/HumanDataCollector/HumanDataCollector.cpp

```diff
@@ -16,18 +16,137 @@
 #include <yarp/os/BufferedPort.h>
 #include <yarp/sig/Vector.h>
 
+#include "matio.h"
+
+// TODO: Double check if the header works on WINDOWS
+#include <experimental/filesystem>
+#include <sys/stat.h>
+
+#include <chrono>
+#include <ratio>
+#include <algorithm>
+#include <utility>
+#include <unordered_map>
+
 const std::string DeviceName = "HumanDataCollector";
 const std::string LogPrefix = DeviceName + " :";
 constexpr double DefaultPeriod = 0.01;
 
 using namespace hde::devices;
 
+void writeVectorOfStringToMatCell(const std::string name, const std::vector<std::string>& strings, mat_t* matFile)
```

Comment:
I'm not fully conviced that you handle the errros properly. I see a lot of `Mat_Close` and no error propagation from these `write*` functions to the outside. You call this method sequentally multiple times around line 850 and if one of those fails, it's not clear how the failure would propagate.

You have few solutions. Among the many, a possible one is returning a bool and use the following approach from the caller's code:

```cpp
if (!writeVectorOfStringToMatCell(...)) {
    return false;
}
```

Then, the string could be passed as reference.

A longer discussion about the raw pointer. It's a bit a personal fight, but in C++ I try to minimize their usage because maintaining them is very demanding and memory problems or null pointer access is more likely to happen. A pattern that I like a lot and used in the past (can't find where) is wrapping the raw pointer inside a `unique_ptr`, defining also a custom destructor. In this way, the object will own the memory and when it goes out of scope, it get deallocated automatically. Furthermore, if this pointer is a file, the handle could be closed automatically as well. This approach will provide a bullet proof error handling since you don't have to mind about who has to free the memory in case of errors, making the propagation way more straighforward.

Reply:
> I'm not fully conviced that you handle the errros properly. I see a lot of Mat_Close and no error propagation from these write* functions to the outside. You call this method sequentally multiple times around line 850 and if one of those fails, it's not clear how the failure would propagate

@diegoferigo the function `writeVectorOfStringToMatCell()` is called only once per data type like `stateJointNames` or `dynamicsJointNames`. The function `writeVectorOfVectorDoubleToMatStruct()` is called only once to fill the mat file with human data stored in the buffers inside `run()`. Both these methods are called in `detach()` call which is done on termination. So, I do not see the need for error propagation from these methods to `detach()` function. At the moment, it is still a preliminary implementation and it will evolve in the future as https://github.com/robotology/human-dynamics-estimation/issues/179#issuecomment-593951539. Given this possible road map, I would like to proceed with this PR and consider the aspects of raw pointer handling (which is mainly related to matio variables) and error propagation (while logging the data through parallel update of buffer variables) as todo items related to matio logging.

Reply:
@diegoferigo https://github.com/robotology/human-dynamics-estimation/pull/181/commits/e1ae3a1d40b8d4ce3dddc0f7f0026307021bf4a4